### PR TITLE
Add form for withdrawing and returning applications

### PR DESCRIPTION
**Pull request by Rhian**
cancel_index
cancel

### Description of change

Created cancel and cancel_confirmation routes which allow all types of user to withdraw or return an application and add a comment. Also added logic for displaying status on the index UI and reinstated user creation to create_sample_data rake task, which was removed earlier in error
### Story Link

https://trello.com/c/Md5k7OGz/2-withdraw-return-applications

### Known issues [OPTIONAL]

We will need to add a ticket for showing on the application page that it has been cancelled, and possibly for displaying the comment.

### CODE REVIEW

File: app/controllers/planning_applications_controller.rb [hunk 1]
```diff
@@ -71,14 +71,37 @@ def withdraw
...
 
+  def cancel_confirmation
+    render :cancel_confirmation
+  end
+
+  def cancel
+    status = authorize_user_can_update_status(
+        planning_application_params[:status])
+    apply_cancellation(status)
+    @planning_application.update!(cancellation_comment: planning_application_params[:cancellation_comment])
```
Comment (james) [inline app/controllers/planning_applications_controller.rb:82]: Could we add the cancellation_comment as an optional parameter to the `withdrawn` and `returned` aasm events? Something like this? https://github.com/aasm/aasm/issues/506#issuecomment-532434609

File: app/controllers/planning_applications_controller.rb [hunk 2]
```diff
@@ -71,14 +71,37 @@ def withdraw
...
+    if @planning_application.withdrawn? || @planning_application.returned?
```
Comment (james) [inline app/controllers/planning_applications_controller.rb:83]: Is there a situation where this wouldn't be true? Do we have a test for the case where it wouldn't be true? If not, I'd remove this `if` statement, as we always expect it to be true.

File: app/controllers/planning_applications_controller.rb [hunk 3]
```diff
@@ -71,14 +71,37 @@ def withdraw
...
+    end
+  end
+
   private
 
   def set_planning_application
     @planning_application = authorize(PlanningApplication.find(params[:id]))
   end
 
   def planning_application_params
```
Comment (james) [inline app/controllers/planning_applications_controller.rb:103]: This has made me realize that we're doing any bulk attribute updating in this controller at all? If that's the case, we can get get rid of this method and it's use of strong parameters.

File: app/views/planning_applications/_assessment_dashboard.html.erb [hunk 4]
```diff
@@ -29,6 +29,11 @@
 </div>
 <div class="govuk-grid-row">
   <div class="govuk-grid-column-two-thirds application">
+    <% unless @planning_application.determined? || @planning_application.returned? || @planning_application.withdrawn? %>
```
Comment (james) [inline app/views/planning_applications/_assessment_dashboard.html.erb:32]: I think we should have a method on PlanningApplication called `cancelable?` or similar for this.

In fact, further to that, I _think_ AASM provides us `@planning_application.may_return?`

File: app/views/planning_applications/_planning_application_table.html.erb [hunk 5]
```diff
@@ -28,16 +32,22 @@
...
         <td class="govuk-table__cell"><%= planning_application.target_date.strftime("%e %b") %></td>
         <td class="govuk-table__cell">
-          <strong class="govuk-tag govuk-tag--<%= days_color(planning_application.days_left) %>">
-            <%= planning_application.days_left %>
-          </strong>
+          <% if id == "closed" %>
+            <strong class="govuk-tag govuk-tag--<%= display_status(planning_application)[:color] %>">
+              <%= display_status(planning_application)[:decision] %>
+            </strong>
+            <% else %>
```
Comment (james) [inline app/views/planning_applications/_planning_application_table.html.erb:39]: The indentation is a bit off here

File: spec/system/planning_applications/cancelling_spec.rb [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+  let!(:policy_evaluation) do
+    create :policy_evaluation,
+            planning_application: planning_application,
+            policy_considerations: [policy_consideration_1, policy_consideration_2]
+  end
+
+  let!(:drawing) do
+    create :drawing, :with_plan, :proposed_tags,
+           planning_application: planning_application
+  end
```
Comment (james) [inline spec/system/planning_applications/cancelling_spec.rb:44]: I think drawing, policy_evaluation, policy_consideration_2 and policy_consideration_1 aren't used or needed for these tests, so we can remove them?

**Review by james — APPROVED** [review]
Apart from all my niggles, functionally this is all good! I'd appreciate the bits of tidying up I've commented, but none are a blocker to merging.

File: app/views/planning_applications/_assessment_dashboard.html.erb (hunk 4 above)
Comment (Rhian) [reply app/views/planning_applications/_assessment_dashboard.html.erb]: I need to investigate further why `.may_return?` and `.may_withdraw?` didn't work as expected. For the moment I've gone with   `true unless determined? || returned? || withdrawn?`

**Pull request merged**